src/anyp/Uri.cc

```diff
@@ -775,8 +835,12 @@ SBuf &
 AnyP::Uri::absolutePath() const
 {
     if (absolutePath_.isEmpty()) {
-        // TODO: Encode each URI subcomponent in path_ as needed.
-        absolutePath_ = Encode(path(), PathChars());
+        if (!path().isEmpty())
+            absolutePath_ = Encode(path(), PathChars());
+        if (!query().isEmpty()) {
+            absolutePath_.append("?");
+            absolutePath_.append(Encode(query(), QueryChars()));
+        }
```

Comment:
FYI: This TODO implementation is the fix for bug 5523 titular issue for PR #2289. The rest of this PR is logic supporting this change.